### PR TITLE
Scenes/VizPanel: Add support for panel repeat options

### PR DESCRIPTION
**Pull request by kaydelaney**
Related scenes changes: https://github.com/grafana/scenes/pull/565

Making this a draft PR as I'm not really sure this is the way to go. Something about modifying the parent's parent's (grandparent's?) state feels "wrong", or clumsy at least. I think most of the friction here comes from having different classes deal with repeated/non-repeated panels, and I've messed around a bit with having a sort of wrapper object that encapsulates the behavior of both, but it's not entirely without its problems either.
In any case, I'd like to hear others' thoughts on what the best way to move forward here is.

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
I wonder if we can have the repeat options on VizPanelManager instead adding options on VizPanel that are not really used / belong there. 

Another thought is to move these options from PanelEdit and instead have a separate modal (or menu actions) for repeat options (Accessible via panel menu).

**Comment by dprokop** [comment]
> I wonder if we can have the repeat options on VizPanelManager instead adding options on VizPanel that are not really used / belong there.

I agree with @torkelo. I think those options make sense only in the edit context while in the view mode they are abstracted away by using different object to represent repeating.

File: public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx [hunk 1]
```diff
@@ -81,8 +83,63 @@ export class PanelEditor extends SceneObjectBase<PanelEditorState> {
...
+          body: panelClone,
+        });
+        sourcePanelGrandparent.setState({
+          children: sourcePanelGrandparent.state.children.map((child) =>
+            child.state.key === sourcePanelParent.state.key ? gridItem : child
+          ),
+        });
+      } else {
+        sourcePanelParent.setState({ source: panelMngr.state.panel.clone() });
+      }
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx:140]: I tested this part and it seems to be working just fine (editing and viewing).

The code looks clumsy, I agree. Faced the same problem with handling multiple variants of changing the data source, but we managed to fix that with structural simplification. I remember trying there an another abpproach by wrapping the objects in yet-another-abstraction to unify the handleing, but that was just delegating complexity to another place.

For now I would focus on trying to get the repeat options into the VizPanelManager as of Torkel's suggestion, and keeping this code as-is.

File: public/app/features/dashboard/components/PanelEditor/getPanelFrameOptions.tsx [hunk 2]
```diff
@@ -261,67 +263,70 @@ export function getPanelFrameCategory2(panel: VizPanel): OptionsPaneCategoryDesc
...
+              return (
+                <Select
+                  options={maxPerRowOptions}
+                  value={panelManager.state.maxPerRow}
+                  onChange={(value) => panelManager.setState({ maxPerRow: value.value })}
+                />
+              );
+            },
+          })
+        )
```
Comment (dprokop) [inline public/app/features/dashboard/components/PanelEditor/getPanelFrameOptions.tsx:330]: Noticed that these options have no effect currently

File: public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx [hunk 3]
```diff
@@ -79,10 +90,63 @@ export class PanelEditor extends SceneObjectBase<PanelEditorState> {
...
+        sourcePanelGrandparent.setState({
+          children: sourcePanelGrandparent.state.children.map((child) =>
+            child.state.key === sourcePanelParent.state.key ? gridItem : child
+          ),
+        });
+      } else {
+        sourcePanelParent.setState({ source: panelManager.state.panel.clone() });
+      }
+    } else {
+      console.error('Unsupported scene object type');
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx:149]: Can we extract this to a separate method and also add some unit tests for this behavior? Quite a lot going on here, would be worth having some coverage.

**Review by harisrozajac — COMMENTED** [review]
Overall this looks good to me. Not sure if this PR is still in draft mode because panel frame options have no effect currently. There's also a merge conflict

File: public/app/features/dashboard/components/PanelEditor/getPanelFrameOptions.tsx [hunk 4]
```diff
@@ -175,7 +176,8 @@ export function getPanelFrameCategory(props: OptionPaneRenderProps): OptionsPane
     );
 }
 
-export function getPanelFrameCategory2(panel: VizPanel): OptionsPaneCategoryDescriptor {
+export function getPanelFrameCategory2(panelManager: VizPanelManager): OptionsPaneCategoryDescriptor {
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply public/app/features/dashboard/components/PanelEditor/getPanelFrameOptions.tsx:179]: We need the panel manager as the repeat options are managed by it. I agree the name `getPanelFrameCategory2` isn't great though

**Comment by kaydelaney** [comment]
> I also noticed this. @kaydelaney will these options be functional in a follow-up PR?

@harisrozajac Do the repeat options not work for you?

https://github.com/grafana/grafana/assets/45561153/a3398ee2-f810-436f-a5e7-559a4bf9d055

**Comment by harisrozajac** [comment]
@kaydelaney the options work! i think i had some mixup with how my variables are configured. tested it and it works nicely now

**Review by harisrozajac — APPROVED** [review]
great work! i think we can move on with this and get back to renaming `getPanelFrameCategory2` and `repeatRowSelect2` later

**Comment by harisrozajac** [comment]
@kaydelaney there's a failing e2e test but i think just resetting the build should fix it. it doesn't seem to be related to any changes in this PR

**Pull request merged**